flatland/envs/agent_chains.py

```diff
@@ -119,99 +117,78 @@ def find_stop_preds(self, svStops=None):
-            svBlocked |= set(lvPred)
-            svBlocked.add(v)
-            # print("node:", v, "set", svBlocked)
-            # only count those not already marked
+            predecessors |= set(lvPred)
+            predecessors.add(v)
+
+            # only color those not already marked (not updating previous colors)
             for v2 in [v] + lvPred:
                 if self.G.nodes[v2].get("color") != color:
                     self.G.nodes[v2]["color"] = color
-                    iCount += 1
-
-        return svBlocked
+        return predecessors
 
     def find_conflicts(self):
-        self.reset_G_reversed()
+        """Called in env.step() before the agents execute their actions."""
 
-        svStops = self.find_stops2()  # voluntarily stopped agents - have self-loops
-        # svSwaps = self.find_swaps()   # deadlocks - adjacent head-on collisions
-        svSwaps = self.find_swaps2()  # faster version of find_swaps
+        svStops: Set[Cell] = self.find_stopped_agents()  # voluntarily stopped agents - have self-loops ("same cell to same cell")
+        svSwaps: Set[Cell] = self.find_swaps()  # deadlocks - adjacent head-on collisions
 
-        # Block all swaps and their tree of predessors
-        self.svDeadlocked = self.block_preds(svSwaps, color="purple")
+        # Mark all swaps and their tree of predecessors with purple - these are directly deadlocked
+        self.svDeadlocked: Set[Cell] = self.mark_preds(svSwaps, color="purple")
 
-        # Take the union of the above, and find all the predecessors
-        # svBlocked = self.find_stop_preds(svStops.union(svSwaps))
-
-        # Just look for the tree of preds for each voluntarily stopped agent
-        svBlocked = self.find_stop_preds(svStops)
+        # Just look for the tree of preds for each voluntarily stopped agent (i.e. not wanting to move)
+        # TODO why not re-use mark_preds(swStops, color="red")?
+        # TODO refactoring suggestion: only one "blocked" red = 1. all deadlocked and their predecessor, 2. all predecessors of self-loopers, 3.
+        svBlocked: Set[Cell] = self.find_stop_preds(svStops)
 
         # iterate the nodes v with their predecessors dPred (dict of nodes->{})
         for (v, dPred) in self.G.pred.items():
-            # mark any swaps with purple - these are directly deadlocked
-            # if v in svSwaps:
-            #    self.G.nodes[v]["color"] = "purple"
-            # If they are not directly deadlocked, but are in the union of stopped + deadlocked
-            # elif v in svBlocked:
 
-            # if in blocked, it will not also be in a swap pred tree, so no need to worry about overwriting
+            dPred: Set[Cell] = dPred
+
+            # if in blocked, it will not also be in a swap pred tree, so no need to worry about overwriting (outdegree always  <= 1!)
+            # TODO why not mark outside of the loop? The loop would then only need to go over nodes with indegree >2 not marked purple or red yet
```

Comment:
yes I can't see any reason why we don't use some sort of bulk action to mark svBlocked as "red"; though it may not save much time as it looks like an O(n) operation.

Reply:
Focus is on readability and not on performance, agreed.